src/backend/distributed/utils/placement_connection.c

```diff
@@ -0,0 +1,483 @@
+	{
+		ConnectionReference *connectionReference =
+			dlist_container(ConnectionReference, placementNode, it.cur);
+		MultiConnection *connection = connectionReference->connection;
+		bool useConnection = false;
+
+		/* use the connection, unless in a state that's not useful for us */
+		if (!connection ||
+			connection->claimedExclusively ||
+			(flags & FORCE_NEW_CONNECTION) != 0 ||
+			returnConnectionReference != NULL)
+		{
+			useConnection = false;
+		}
+		else
+		{
+			useConnection = true;
+		}
+
+		/*
+		 * If not using the connection, verify that FOR_DML/DDL flags are
+		 * compatible.
+		 */
+		if (useConnection)
+		{
+			returnConnectionReference = connectionReference;
+		}
+		else if (connectionReference->hadDDL)
+		{
+			/* XXX: errcode & errmsg */
+			ereport(ERROR, (errcode(ERRCODE_ACTIVE_SQL_TRANSACTION),
+							errmsg("cannot establish new placement connection when other "
+								   "placement executed DDL")));
+		}
+		else if (connectionReference->hadDML)
+		{
+			/* XXX: errcode & errmsg */
+			ereport(ERROR, (errcode(ERRCODE_ACTIVE_SQL_TRANSACTION),
+							errmsg("cannot establish new placement connection when other "
+								   "placement executed DML")));
+		}
+	}
+
+	/*
+	 * Either no caching desired, or no connection present. Start connection
+	 * establishment.
+	 */
+	if (returnConnectionReference == NULL)
+	{
+		MultiConnection *connection = StartNodeConnection(flags, placement->nodeName,
+														  placement->nodePort);
+
+		/* FIXME: use xact context ? */
+		returnConnectionReference = (ConnectionReference *)
+									MemoryContextAlloc(ConnectionContext,
+													   sizeof(ConnectionReference));
+		returnConnectionReference->connection = connection;
+		returnConnectionReference->hadDDL = false;
+		returnConnectionReference->hadDML = false;
+		returnConnectionReference->shardId = placement->shardId;
+		returnConnectionReference->placementId = placement->placementId;
+		dlist_push_tail(&placementEntry->connectionReferences,
+						&returnConnectionReference->placementNode);
+
+		/* record association with shard, for invalidation */
+		AssociatePlacementWithShard(placementEntry, placement);
+
+		/* record association with connection, to handle connection closure */
+		dlist_push_tail(&connection->referencedPlacements,
+						&returnConnectionReference->connectionNode);
+	}
+
+	if (flags & FOR_DDL)
+	{
+		returnConnectionReference->hadDDL = true;
+	}
+	if (flags & FOR_DML)
+	{
+		returnConnectionReference->hadDML = true;
+	}
+
+	return returnConnectionReference->connection;
+}
+
+
+/* Record shard->placement relation */
+static void
+AssociatePlacementWithShard(ConnectionPlacementHashEntry *placementEntry,
+							ShardPlacement *placement)
+{
+	ConnectionShardHashKey shardKey;
+	ConnectionShardHashEntry *shardEntry = NULL;
+	bool found = false;
+	dlist_iter placementIter;
+
+	shardKey.shardId = placement->shardId;
+	shardEntry = hash_search(ConnectionShardHash, &shardKey, HASH_ENTER, &found);
```

Comment:
To record an in-person conversation Marco and I just had:
We decided to punt on this for now. The current hack of using ```ColocatedShardIntervalList(LowestShardIntervalById())``` will continue to work, and a proper solution isn't entirely trivial. Since this PR is blocking others, it seems better to tackle that issue separately.